IGNITE-24976 Sql. Optimize sql row serialization

AMashenkov opened the pull request:
Thank you for submitting the pull request.

To streamline the review process of the patch and ensure better code quality
we ask both an author and a reviewer to verify the following:

### The Review Checklist
- [ ] **Formal criteria:** TC status, codestyle, mandatory documentation. Also make sure to complete the following:  
\- There is a single JIRA ticket related to the pull request.  
\- The web-link to the pull request is attached to the JIRA ticket.  
\- The JIRA ticket has the Patch Available state.  
\- The description of the JIRA ticket explains WHAT was made, WHY and HOW.  
\- The pull request title is treated as the final commit message. The following pattern must be used: IGNITE-XXXX Change summary where XXXX - number of JIRA issue.
- [ ] **Design:** new code conforms with the design principles of the components it is added to.
- [ ] **Patch quality:** patch cannot be split into smaller pieces, its size must be reasonable.
- [ ] **Code quality:** code is clean and readable, necessary developer documentation is added if needed.
- [ ] **Tests code quality:** test set covers positive/negative scenarios, happy/edge cases. Tests are effective in terms of execution time and resources.

### Notes
- [Apache Ignite Coding Guidelines](https://cwiki.apache.org/confluence/display/IGNITE/Java+Code+Style+Guide)

korlov42 reviewed (commented):
I see we eliminated boxing, but I would suggest to go even further: we can take advantage of `org.apache.ignite.internal.binarytuple.BinaryTupleBuilder#appendElementBytes` for columns which actually presented in the underlying materialized tuple from store. This way we will eliminate not only boxing, but also deserialization of non-primitive types. WDYT?

AMashenkov commented:
I've added a UpgradeRowBenchmark that runs simple select with and without projection.

```
Master (no row upgrading)
Benchmark                               (clusterSize)  (fsync)  Mode  Cnt    Score   Error  Units
UpgradeRowBenchmark.selectAll                       1    false  avgt   20  100.949 ± 0.178  ms/op
UpgradeRowBenchmark.selectAllProjected              1    false  avgt   20  137.263 ± 0.319  ms/op

My branch (no row upgrading)
Benchmark                               (clusterSize)  (fsync)  Mode  Cnt    Score   Error  Units
UpgradeRowBenchmark.selectAll                       1    false  avgt   20  102.289 ± 0.144  ms/op
UpgradeRowBenchmark.selectAllProjected              1    false  avgt   20  131.582 ± 0.136  ms/op
```

The same benchmark, but i've added a column after populating the table.
```
Master (with row upgrading)
Benchmark                               (clusterSize)  (fsync)  Mode  Cnt    Score   Error  Units
UpgradeRowBenchmark.selectAll                       1    false  avgt   20  294.983 ± 0.785  ms/op    <--- this looks unexpected
UpgradeRowBenchmark.selectAllProjected              1    false  avgt   20  189.532 ± 0.213  ms/op

My branch (with row upgrading)
Benchmark                               (clusterSize)  (fsync)  Mode  Cnt    Score   Error  Units
UpgradeRowBenchmark.selectAll                       1    false  avgt   20  199.317 ± 0.403  ms/op
UpgradeRowBenchmark.selectAllProjected              1    false  avgt   20  177.769 ± 0.274  ms/op
```

korlov42 commented on modules/runner/src/integrationTest/java/org/apache/ignite/internal/benchmark/UpgradeRowBenchmark.java:
```diff
@@ -0,0 +1,115 @@
+    }
+}
+
+
```
extra line

korlov42 commented on modules/binary-tuple/src/main/java/org/apache/ignite/internal/binarytuple/BinaryTupleParser.java:
```diff
@@ -504,6 +504,22 @@ public final Period periodValue(int begin, int end) {
+     * @param index Index of the element.
+     */
+    protected void copyRawValue(BinaryTupleBuilder builder, int index) {
+        fetch(index, (idx, begin, end) -> {
```
I think we can avoid creation of lambda on every read by moving this method to `BinaryTupleReader`. As downside we will need to make buffer package-private. WDYT?

korlov42 commented on modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/util/ProjectedTuple.java:
```diff
@@ -17,26 +17,22 @@
  * @see AbstractProjectedTuple
  */
-public class FormatAwareProjectedTuple extends AbstractProjectedTuple {
+public class ProjectedTuple extends AbstractProjectedTuple {
```
probably we should merge `ProjectedTuple` and `AbstractProjectedTuple` altogether. So we will have only `ProjectedTuple` and extension which append virtual columns. WDYT?

korlov42 commented on modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/util/ProjectedTuple.java:
```diff
@@ -17,26 +17,22 @@
- * (de-)serialization during tuple normalization.
+ * A projected tuple wrapper that is best effort to avoiding unnecessary (de-)serialization during tuple normalization.
  *
  * <p>It's up to the caller to get sure that provided tuple respect the format.
```
this line is no longer valid, let's remove it

korlov42 commented on modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/util/ProjectedTuple.java:
```diff
@@ -45,49 +41,47 @@ public class FormatAwareProjectedTuple extends AbstractProjectedTuple {
-    public FormatAwareProjectedTuple(InternalTuple delegate, int[] projection) {
-        super(delegate, projection);
+    public ProjectedTuple(InternalTuple delegate, int[] projection) {
+        super((InternalTupleEx) delegate, projection);
```
tbh I don't really like this approach with force casting. Let's try to preserve type safety

AMashenkov replied:
Fixed
